src/api/source/TileClient.js

```diff
@@ -0,0 +1,80 @@
+import { validateTemplateURL } from '../url';
+import DataframeCache from './DataframeCache';
+import { rTiles } from '../../client/rsys';
+
+
+export default class TileClient {
+    constructor(templateURLs) {
+        if (!Array.isArray(templateURLs)) {
+            templateURLs = [templateURLs];
+        }
+        templateURLs.map(validateTemplateURL);
+        this._templateURLs = templateURLs;
+        this._requestGroupID = 0;
+        this._oldDataframes = [];
+        this._cache = new DataframeCache();
+    }
+
+    bindLayer(addDataframe, dataLoadedCallback) {
+        this._addDataframe = addDataframe;
+        this._dataLoadedCallback = dataLoadedCallback;
+    }
+
+    requestData(viewport, responseToDataframeTransformer) {
+        const tiles = rTiles(viewport);
+        this._getTiles(tiles, responseToDataframeTransformer);
+    }
+
+    free() {
+        this._cache = new DataframeCache();
+        this._oldDataframes = [];
+    }
+
+    _getTiles(tiles, responseToDataframeTransformer) {
+        this._requestGroupID++;
+        let completedTiles = [];
+        let needToComplete = tiles.length;
+        const requestGroupID = this._requestGroupID;
+        tiles.forEach(t => {
+            const { x, y, z } = t;
+            this._cache.get(`${x},${y},${z}`, () => this._requestDataframe(x, y, z, responseToDataframeTransformer)).then(
```

Comment:
this is not working as a cache and IMO is very confusing.

IMO the fetch logic should remain in the MVT client and the cache must remain a simple and classic cache!

something like:
```js
if (this._cache.has(ID)){
   this._cache.get(ID).then(this._onDataframeCallback)
}
else {
   const dataFramePromise = this._requestDataframe(x, y, z, responseToDataframeTransformer);
   this.cache.set(ID, dataframePromise);
   dataframePromise.then(this._onDataframeCallback);
}


```

Reply:
The current cache is uncoupled with the tile format, what you are proposing is to duplicate the cache code for each format. For a common cache class I already have LRU-cache, what I need is another layer that encapsulates the dataframe-related logic of the cache.

Reply:
> The current cache is uncoupled with the tile format, what you are proposing is to duplicate the cache code for each format

What I am proposing is even less coupled since the cache simply stores promises.

> for each format

Which formats do we have? 

> What I need is another layer that encapsulates the dataframe-related logic of the cache.

I dont follow you

Reply:
Of course, the code you still leave at the cache is uncoupled, but you move code to the MVT file, that code is coupled.

Before following refactors the DataframeCache class was shared by the MVT source and the Windshaft source. After more refactoring, windshaft just use MVT, this means that the DataframeCache is used right now by only one class (MVT), but the abstraction worked, works, and if we add a new format in the future it will still work.

The DataframeCache encapsulates the logic of "how many dataframes to keep", "for how long", and "what to do in case we need to remove one from the cache". To do it, it needs to know about Dataframes, it doesn't need to know about MVT/GeoJSON/...

Reply:
> but you move code to the MVT file, that code is coupled.

which code? all the code was already there 

> Before following refactors the DataframeCache class was shared by the MVT source and the Windshaft source. After more refactoring, windshaft just use MVT, this means that the DataframeCache is used right now by only one class (MVT), but the abstraction worked, works, and if we add a new format in the future it will still work.

In which way my proposal is not compatible with that?


> The DataframeCache encapsulates the logic of "how many dataframes to keep", "for how long", and "what to do in case we need to remove one from the cache". To do it, it needs to know about Dataframes, it doesn't need to know about MVT/GeoJSON/...

With the current approach the cache knows nothing about the dataframe, but needs a second parameter in the `get` function I never seen before in a cache

Reply:
eing?